pkg/restore/cdclog/decoder.go

```diff
@@ -0,0 +1,273 @@
+				log.Fatal("invalid column value, please report a bug", zap.Any("col", c), zap.Error(err))
+			}
+			c.Value = uint64(intNum)
+		}
+	}
+	return c
+}
+
+type messageKey struct {
+	Ts        uint64 `json:"ts"`
+	Schema    string `json:"scm,omitempty"`
+	Table     string `json:"tbl,omitempty"`
+	RowID     int64  `json:"rid,omitempty"`
+	Partition *int64 `json:"ptn,omitempty"`
+}
+
+// Encode the messageKey.
+func (m *messageKey) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the messageKey.
+func (m *messageKey) Decode(data []byte) error {
+	return json.Unmarshal(data, m)
+}
+
+// MessageDDL represents the ddl changes.
+type MessageDDL struct {
+	Query string             `json:"q"`
+	Type  timodel.ActionType `json:"t"`
+}
+
+// Encode the DDL message.
+func (m *MessageDDL) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the DDL message.
+func (m *MessageDDL) Decode(data []byte) error {
+	return json.Unmarshal(data, m)
+}
+
+// MessageRow represents the row changes in same commit ts.
+type MessageRow struct {
+	Update     map[string]column `json:"u,omitempty"`
+	PreColumns map[string]column `json:"p,omitempty"`
+	Delete     map[string]column `json:"d,omitempty"`
+}
+
+// Encode the Row message.
+func (m *MessageRow) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the Row message.
+func (m *MessageRow) Decode(data []byte) error {
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.UseNumber()
+	err := decoder.Decode(m)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for colName, column := range m.Update {
+		m.Update[colName] = formatColumnVal(column)
+	}
+	for colName, column := range m.Delete {
+		m.Delete[colName] = formatColumnVal(column)
+	}
+	for colName, column := range m.PreColumns {
+		m.PreColumns[colName] = formatColumnVal(column)
+	}
+	return nil
+}
+
+// SortItem represents a DDL item or Row changed item.
+type SortItem struct {
+	ItemType ItemType
+	Data     interface{}
+	Schema   string
+	Table    string
+	RowID    int64
+	TS       uint64
+}
+
+// LessThan return whether it has smaller commit ts than other item.
+func (s *SortItem) LessThan(other *SortItem) bool {
+	if other != nil {
+		return s.TS < other.TS
+	}
+	return false
+}
+
+// JSONEventBatchMixedDecoder decodes the byte of a batch into the original messages.
+type JSONEventBatchMixedDecoder struct {
+	mixedBytes []byte
+}
+
+func (b *JSONEventBatchMixedDecoder) decodeNextKey() (*messageKey, uint64, error) {
+	keyLen := binary.BigEndian.Uint64(b.mixedBytes[:8])
+	key := b.mixedBytes[8 : keyLen+8]
+	// drop value bytes
+	msgKey := new(messageKey)
+	err := msgKey.Decode(key)
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+	return msgKey, keyLen, nil
+}
+
+// NextEvent return next item depends on type.
+func (b *JSONEventBatchMixedDecoder) NextEvent(itemType ItemType) (*SortItem, error) {
+	if !b.HasNext() {
+		return nil, nil
+	}
+	nextKey, nextKeyLen, err := b.decodeNextKey()
```

Comment:
the nextKeyLen is only used here, so we can directly set b.mixedBytes = b.mixedBytes[nextKeyLen+8:] in decodeNextKey, then we needn't return the nextKeyLen

Reply:
Done, PTAL